src/System.Management.Automation/engine/Modules/TestModuleManifestCommand.cs

```diff
@@ -145,15 +145,20 @@ protected override void ProcessRecord()
                             }
                         }
 
-                        // RootModule can be null, empty string or point to a valid .psm1, , .cdxml, .xaml or .dll.  Anything else is invalid.
+                        /*
+                            RootModule can be null, empty string or point to a valid .psm1 (extension is optional), .cdxml, .xaml or .dll.
+                            Anything else is invalid.
+                         */
                         if (module.RootModule != null && module.RootModule != string.Empty)
                         {
                             string rootModuleExt = System.IO.Path.GetExtension(module.RootModule);
-                            if ((!IsValidFilePath(module.RootModule, module, true) && !IsValidGacAssembly(module.RootModule)) ||
+
+                            if ((!rootModuleExt.Equals(string.Empty) && !IsValidFilePath(module.RootModule + ".psm1", module, true)) &&
```

Comment:
Actually, this entire condition really needs to be spun into a `bool` method. I think it's now too long to be inline here, since it's very hard to read and determine the logic or intention of.

Something like:

```csharp
...
if (!IsRootModuleValid(module.RootModule))
{
    string errorMsg = StringUtil.Format(Modules.InvalidModuleManifest, module.RootModule, filePath);
    ...
}
...

/// <summary>
/// Checks whether the RootModule field of a module is valid or not.
/// Valid root modules are:
///  - null
///  - Empty string
///  - A valid non-psd1 module file (psm1, cdxml, xaml, dll), as name with extension, name without extension, or path
/// </summary>
/// <param name="rootModule">The root module specified in the module manifest RootModule field.</param>
/// <returns>True if the root module is valid, false otherwise.</returns>
private static bool HasValidRootModule(PSModuleInfo module)
{
    // Empty/null root modules are allowed
    if (string.IsNullOrEmpty(module.RootModule))
    {
        return true;
    }

    // If the root module has an extension, it may only be a known module extension
    string rootModuleExt = Path.GetExtension(module.RootModule);
    if (!string.IsNullOrEmpty(rootModuleExt) && !ModuleIntrinsics.PSModuleExtensions.Contains(rootModuleExt, StringComparison.OrdinalIgnoreCase))
    {
        return false;
    }

    // GAC assemblies are allowed
    if (IsValidGacAssembly(module.RootModule))
    {
        return true;
    }

    // We have a valid file name, but must check the path
    return IsValidFilePath(module.RootModule, module, verifyPathScope: true);
}
```

I haven't executed or otherwise tested this code, so it might not even compile, but I think it's worth breaking the logic up in this PR.

Reply:
Thanks for the advice. From what I understand a root module without an extension is considered valid only if the file path with extension is valid and that means we can't just return true if the root module has no file extension as this code suggest.

Reply:
Yes, that makes sense, good catch.

I've edited the code above. Tell me if that makes more sense.

Reply:
Yep. I'll rewrite the PR on monday using this. Thanks for the help. :)